### ISSUE TITLE
Trying to reproduce the results of ZESRec

### ISSUE THREAD

**Issue by xukp20** [RUCAIBox/UniSRec#10]
Thanks for your great work! I'm trying to reproduce the results of ZESRec from your paper and I'm not sure about the settings. 
Could you provide some details about how can I implement the ZESRec model based on UniSRec's code or SASRec in Recbole, and it would be nice if you could also provide the settings of the pretrain/finetune phase. Thank you!

**Comment by xukp20** [comment]
I have add the Bert model to the front and use the two layer MLP (uen in ZESRec paper) to map the dimension to 300 for SASRec, then I remove the GRU architecture in the original paper and replace it with SASRec( just by inheriting the SASRec model), the output of SASRec part would be the user's embedding. I use cross entropy to measure the propability of the target item among all the items in the dataset. Am I doing right for the model structure? And how about the time/GPUs it needs to pretrain the model on the FHCKM dataset? Thank you very much!

**Comment by hyp1231** [comment]
Thanks for your attention! I believe that most of your ZESRec implementations are the same as mine.

Note that (1) the BERT model is fixed, as a result, we preprocess the BERT representations of each item. In this way, ZESRec actually shares a same input as UniSRec. (2) We found that the performance of ZESRec drops significantly if we have non-linear activation layers in "the two layer MLP". To reproduce the results of ZESRec, I only use a single (768 * 300) linear layer as the MLP.

I also share some implementation suggestions in #4 .
> For the baseline ZESRec, to the best of our knowledge, the official implementation of ZESRec has not been released, thus we prefer not to upload the corresponding code recently. But I believe that it can be implemented in an easy way. One similar implementation can be regarded as the ablation of UniSRec, by replacing the MoE-enhanced adaptor to one linear module, and removing additional pre-training tasks.

I pre-trained ZESRec on FHCKM dataset with 4 * 24G 3090 GPUs for <2 days. Hope these can help more or less.

**Comment by xukp20** [comment]
thanks a lot! I just realized I've implemented the 'uen' according to the version 1 of Zesrec，and the new version seems to use one linear layer just as you said :）

**Comment by xukp20** [comment]
Sorry but I still have some questions: 1. I use the cross entropy loss in my model,  but the v2 paper of Zesrec added two L2 regulation terms, should I added these too? 2. If bert is fixed when pretraining, I think I should modify the bert weight and also uen, sasrec part when fine tuning? 3. how should I set the training hyper param terse like epoch number, is it fine to use  the same as Unisrec pretraining? Thanks again!

**Comment by hyp1231** [comment]
> Sorry but I still have some questions: 1. I use the cross entropy loss in my model, but the v2 paper of Zesrec added two L2 regulation terms, should I add these too? 2. If bert is fixed when pretraining, I think I should modify the bert weight and also uen, sasrec part when fine tuning? 3. how should I set the training hyper paramters like epoch number, is it fine to use the same as Unisrec pretraining? Thanks again!

No worries!

1. I once tried to add some while reproducing ZESRec's own results. It could be my problems but I really cannot get better performance from adding L2 regulation terms. As a result, I didn't add L2 regulation terms while pre-training on FHCKM.
2. In my implementation, I also fixed the parameters of BERT model, while others are all tuned to make a fair comparison.
3. I pre-trained ZESRec for 300 epochs to make a fair comparison.

**Comment by xukp20** [comment]
ok, that's clear enough for me, thanks a lot 🤣

**Comment by xukp20** [comment]
Hello there and it's me again. I have done with pretraining for ZESRec and I am trying to figure out the settings for the finetuning phase. I have known that the bert part should be fixed but what about the linear(uen) and the SASRec part? I think the SASRec is fixed by default according to the finetune.py script as the trm_encoder&positional_embedding require no grad, is that right? 
I'm running the script using this command
```shell
python finetune.py -d <down_stream_dataset> -p <my zesrec pth> --train_stage=inductive_ft
```
I have changed the finetune.py to init ZESRec model instead of UniSRec, please tell me if anything is wrong compared to your settings.
Sorry for bothering you again. Thanks a lot : )

**Comment by hyp1231** [comment]
Hi, no worries! :)

In my implementation, I do not fix the parameters of the linear and SASRec parts. The reason is that the original paper does not discuss some parameter-efficient fine-tuning tricks, and I'm afraid that fixing the parameters of SASRec part (and uen part) may lead to a performance drop.

**Comment by XiaoLongtaoo** [comment]
@xukp20 


> Sorry but I still have some questions: 1. I use the cross entropy loss in my model, but the v2 paper of Zesrec added two L2 regulation terms, should I add these too? 2. If bert is fixed when pretraining, I think I should modify the bert weight and also uen, sasrec part when fine tuning? 3. how should I set the training hyper paramters like epoch number, is it fine to use the same as Unisrec pretraining? Thanks again!

Hello, I'm also trying to reproduce ZESRec. I want to know the code difference between UniSRec and ZESRec, could u provide the config file and code with me? Thanks a lot!
